User entitlements workshop (#1615)
oktadeveloper/okta-blog: _source/_posts/2026-01-21-user-entitlements-workshop.md

  * Enable **Create Users**
  * Enable **Update User Attributes**
  * Enable **Deactivate Users**
  
Press **Save**. 

Import users from the todo app into Okta. Navigate to the **Import** tab and press the **Import Now** button. Okta discovers users in your app and tries to match them with users already defined in Okta. A dialog shows Okta discovered the two users you added using the DB script. Select both users and press the **Confirm Assignments** to confirm the assignments.

You'll see the imported users in the **Assignments** tab. But what about entitlements? They're coming right up!

Stop the tunnel and the API using the <kbd>Ctrl</kbd>+<kbd>c</kbd> command in the terminal windows. We'll make some changes to the API that won't automatically reflect in the local tunnel, so we'll get all our entitlements changes made and resynchronize with Okta.

## SCIM schemas and resources 

In the first SCIM workshop, you learned about SCIM's `User` resource and built out operations around the user. You updated only a handful of user properties in the workshop, but SCIM is way more powerful thanks to its superpower—_extensibility_. ✨ User is not the only resource type defined in SCIM.

> A `Resource` represents an object SCIM operates on, such as a user or group. SCIM identified core properties each `Resource` must define, such as `id` and a link to the resource's schema definition. From there, a user extends from the core properties and adds attributes specific to the object, such as adding `userName` and their emails. A standard published schema exists for all those user-specific attributes within the SCIM spec. You can continue extending resources as needed to represent new resources, such as another SCIM standard-defined schema for Enterprise User. 

Reviewer: ```suggestion
A `Resource` represents an object SCIM operates on, such as a user or group. SCIM identifies [core properties](https://datatracker.ietf.org/doc/html/rfc7643#section-3.1) each `Resource` must define, such as `id` and a link to the resource's [schema definition](https://datatracker.ietf.org/doc/html/rfc7643#section-7). From there, a user extends from the core properties and adds attributes specific to the object, such as adding `userName` and their emails. A standard published [schema](https://datatracker.ietf.org/doc/html/rfc7643#section-4.1) exists for all those user-specific attributes within the SCIM spec. You can continue extending resources as needed to represent new resources, such as another SCIM standard-defined schema for Enterprise User. 
```
'identified' in the past perfect invites the question "when did it stop?", whereas matching present progressive with adjacent sentences reads more smoothly. also I am the kind of reader who sees a few example properties and wonders what the rest are. 

I'm not certain that these are the best possible links to convey the intended meaning, but the paragraph makes more sense to me with them than without.

{% img blog/user-entitlements-workshop/resource-class-diagram.svg alt:"Class diagram representing core Resource properties, User class extending from core Resource adds username and emails properties. The Enterprise User class extends from User adds department and costCenter properties. Group class extends from core Resource and adds displayName and members properties. Other class extends from core Resource demonstrating new resource representations." width:"800" %}{: .center-image }

{% comment %}
Tweak the diagram on https://mermaid.live/ with the following content
classDiagram
    Resource <|-- User
    Resource <|-- Group
    User <|-- Enterprise User
    Resource <|-- Other
    Resource : id
    Resource: schemas
    User : userName
    User : list of emails
    Group: list of members
    EnterpriseUser : department
    EnterpriseUser : costCenter
{% endcomment %}

> What's an example resource other than a user or group? If you said "role" or an "entitlement," you're correct! Those resource types must have an `id` and `schemas`. Here, Okta used SCIM's extensibility to define a new resource type.

Reviewer: ```suggestion
What's an example resource other than a user or group? If you said "role" or "entitlement," you're correct! Those resource types must have an `id` and `schemas`. Here, Okta used SCIM's extensibility to define a new resource type.
```
or equivalently, "a role or an entitlement". I had to look up whether this is technically a real grammar rule or just an aesthetic preference, and learned about https://advice.writing.utoronto.ca/revising/faulty-parallelism/

{% img blog/user-entitlements-workshop/resource-role-class-diagram.svg alt:"Class diagram representing core Resource properties. The User, Group, OktaRole, and Other class extends from core Resource." width:"800" %}{: .center-image }

{% comment %}
Tweak the diagram on https://mermaid.live/ with the following content
classDiagram
    Resource <|-- User
    Resource <|-- Group
    Resource <|-- OktaRole
    Resource <|-- Other
    Resource : id
    Resource: schemas
{% endcomment %}

> Okta defines a schema for the `Role` representation. We can use the schema to ensure we conform to the definition.

Reviewer: ```suggestion
Okta defines a [schema](https://developer.okta.com/docs/guides/scim-with-entitlements/main/#role-example) for the `Role` representation. We can use the schema to ensure we conform to the definition.
```

maybe there's a better link? if so, it has dismal SEO. Either way, showing the schema we mean is nice.

### Harness TypeScript to conform to SCIM schemas

We can define an interface to model the `Role` representation. Add a new file to the project named `okta-enterprise-ready-workshops/apps/api/src/scim-types.ts` and open it up in the IDE. This file will contain the SCIM schema definitions, such as the SCIM core `Resource`. Each interface defines required and optional properties and the property's type.

Copy and paste the first interface for the SCIM resource into the `scim-types.ts` file.

```ts
export interface IScimResource {
  id: string;
  schemas: string[];
  meta?: IMetadata;
}
```

A SCIM resource has an optional meta property containing the resource's metadata. Your IDE shows errors, so we can fix this by adding the `IMetadata` definition to the file below the `IScimResource`:

```ts
export interface IMetadata {
  resourceType: RESOURCE_TYPES;
  location?: string;
}
```

You'll have a new error for `RESOURCE_TYPES`. We'll fix it soon.

> Now, on to the Okta `Role` representation. The role representation extends from the core SCIM resource and adds extra properties. Okta's schema overlaps with the SCIM standard User `roles` field, which includes a property for `display` text. Define the interface and add it to `IMetadata` below.

Reviewer: ```suggestion
Now, on to the Okta `Role` representation. The role representation extends from the core SCIM resource and adds extra properties. Okta's schema overlaps with the SCIM standard User `roles` field, which includes a property for `display` text. Define the interface and add it below `IMetadata`.
```

i think that this is what you mean because it describes how the finished code looks in https://github.com/oktadev/okta-enterprise-ready-workshops/blob/entitlements-workshop-complete/apps/api/src/scim-types.ts

```ts
export interface IOktaRole extends IScimResource{
  displayName: string;
}
```

> The `IOktaRole` extends from the core `IScimResource` interface and adds a new required property, `displayName`. Each resource requires a schema, a Uniform Resource Namespace (URN) string. Instead of repeatedly typing the string for each role resource, define it below the `IOktaRole` interface for reusability

Reviewer: ```suggestion
The `IOktaRole` extends from the core `IScimResource` interface and adds a new required property, `displayName`. Each resource requires a schema, a Uniform Resource Namespace (URN) string. Instead of repeatedly typing the string for each role resource, define it below the `IOktaRole` interface for reusability:
```
all the other paragraphs end with punctuation so this one was probably meant to as well?

```ts
export const SCHEMA_OKTA_ROLE = 'urn:okta:scim:schemas:core:1.0:Role';
```

Let's fix the `RESOURCE_TYPES` error. Below the `SCHEMA_OKTA_ROLE` constant, add the following:

```ts
export type RESOURCE_TYPES = 'Role';
```

You can use the `IOktaRole` interface in the `/Roles` endpoint to ensure the response matches the expected structure. Open `okta-enterprise-ready-workshops/apps/api/src/entitlements.ts`, and update the code to use the interface.

```ts
import { Router } from 'express';
import { IOktaRole, SCHEMA_OKTA_ROLE } from './scim-types';

export const rolesRoute = Router();

rolesRoute.route('/')
.get(async (req, res) => {
  const roles: IOktaRole[] = [{
    schemas: [SCHEMA_OKTA_ROLE],
    id: 'one',
    displayName: 'Todo-er'
  }];

  return res.json(roles);
});
```

> **Why use TypeScript and interfaces?**
> 

> > TypeScript, a superset of JavaScript, supports type safety. Type safety means we'll catch errors within the IDE or at build time instead of getting caught by surprise with a runtime error. Here, we state the `roles` array is of type `IOktaRole[]`. Try commenting out the required `schemas` property. You'll see an error in an IDE that supports TypeScript or when you try to serve the API as console output. We can use type safety to ensure we meet the expectations of required SCIM properties in our calls.

Reviewer: this is a very well-phrased, clear, and helpful note.

>
> {% img blog/user-entitlements-workshop/type-error.jpg alt:"IDE and terminal showing the type error when `schemas` is commented out" width:"600" %}{: .center-image }

Every code change deserves a quick check. Serve the API and double check everything still works for you when you make the HTTP call to

```http
GET http://localhost:3333/scim/v2/Roles HTTP/1.1
```

Do you see the one 'Todo-er' role in the response? ✅

### SCIM list response 

> We return the array of Okta roles directly in the API response, but this format doesn't match SCIM list responses. SCIM has a structured response format for lists and a defined schema. This way, SCIM structures all communication between the client and the server so each side knows how to format and parse data. 

Reviewer: ```suggestion
We return the array of Okta roles directly in the API response, but this format doesn't match SCIM list responses. SCIM has a structured response format for lists and a defined [schema](https://datatracker.ietf.org/doc/html/rfc7644#section-3.4.2). This way, SCIM structures all communication between the client and the server so each side knows how to format and parse data. 
```

again, not sure if this is the best link, but it helps some

    schemas: [SCHEMA_LIST_RESPONSE],
    totalResults: roles.length,
    startIndex,
    itemsPerPage: recordLimit,
    Resources: roles.map(role => ({
      schemas: [SCHEMA_OKTA_ROLE],
      id: role.id.toString(),
      displayName: role.name
    }))
  };

  return res.json(listResponse);
});
```

Run a quick check to ensure everything still works. Serve the API and call the `/Roles` endpoint using your HTTP client. ✅

You should see three roles matching the roles in the database. 🎉

## SCIM resource types 

We implemented the `/Roles` endpoint and discussed how SCIM defines a resource. But how would the SCIM client know about this Okta Role type? Enter discovery—learning about a SCIM server's capabilities and supported objects such as resources!

SCIM clients and servers communicate about the types of resources through a standard endpoint, the`/ResourceType` endpoint. SCIM clients call the endpoint to discover what resources they can expect. The endpoint returns a SCIM list response outlining resources. You can add every resource type used, including the standard `User` and `EnterpriseUser` resources, but Okta expects resource definitions only for custom types. 

> First, we'll create the interface for the `ResourceType` and define some strings. Open `okta-enterprise-ready-workshops/apps/api/src/scim-types.ts`. Add the interface for `IResourceType` above the `IListResponse` interface.

Reviewer: other steps have specified "add it below", and this one switches to "add it above". I do not know typescript well enough to feel sure of whether this change is for a technical reason or just to keep variety in the flow of the writing. Not calling anything incorrect here, just noting that there's a moment of "I wonder why that changed" in the reading experience.

```ts
export interface IResourceType {
  id?: string;
  schemas: string[];
  name: string; 
  description?: string;
  endpoint: string;
  schema: string; 
  meta: IMetadata;
}
```

> Notice the `IResourceType` doesn't extend from the `IScimResource` interface. For example, the SCIM standard doesn't require `id` for a resource type. Since the SCIM standard treats `ResourceType` as an exception case of `Resource`, we defined it separately without the relation instead of extending from `IScimResource`.

Reviewer: this is very helpful to have spelled out, and resolves the confusion I experienced back up in `## SCIM schemas and resources`

When following the SCIM protocol, responses that list values, such as the list of roles or resource types, use the SCIM list response format.

The `IListResource` interface must support `IOktaRole` and `IResourceType`. Using [generics](https://www.typescriptlang.org/docs/handbook/2/generics.html) and [union types](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#union-types), we can support different list response objects . Update the `IListResource` to match the code below.

```ts
export interface IListResponse<T extends IScimResource | IResourceType> {
  schemas: string[];
  totalResults: number;
  startIndex: number;
  itemsPerPage: number;
  Resources: T[];
}
```

> You'll see errors in the IDE and, if you're running the API, within the console output. No worries; we'll fix those errors soon!

Reviewer: somewhere around here, an error showed up where prisma suddenly forgot that we'd just taught it about roles: 
<img width="947" height="120" alt="image" src="https://github.com/user-attachments/assets/ee438e43-c259-4102-afcc-83e18c1c9171" />

It went away when I ran `TypeScript: Restart TS Server` in the vscode command palette. This was probably just a random problem with my vscode; only if others end up troubleshooting the same thing would it be worth mentioning in the post.

Resource types have a schema URN and use "ResourceType" as the `resourceType` string in the metadata. Add `SCHEMA_RESOURCE_TYPE` and edit `RESOURCE_TYPES` so your string constants section looks like the code below.

```ts
export const SCHEMA_OKTA_ROLE = 'urn:okta:scim:schemas:core:1.0:Role';
export const SCHEMA_LIST_RESPONSE = 'urn:ietf:params:scim:api:messages:2.0:ListResponse';
export const SCHEMA_RESOURCE_TYPE = 'urn:ietf:params:scim:schemas:core:2.0:ResourceType';
export type RESOURCE_TYPES = 'Role' | 'ResourceType';
```

Open `okta-enterprise-ready-workshops/apps/api/src/entitlements.ts`. Let's fix the `IListResponse` error for the `/Roles` endpoint and specify the object type in the list, the `IOktaRole` type. The code building out the list changes to

```ts
  const listResponse: IListResponse<IOktaRole> = {
    schemas: [SCHEMA_LIST_RESPONSE],
    totalResults: roles.length,
    startIndex,
    itemsPerPage: recordLimit,
    Resources: roles.map(role => ({
      schemas: [SCHEMA_OKTA_ROLE],
      id: role.id.toString(),
      displayName: role.name
    }))
  };
```

You shouldn't see errors anymore! 🎉

> We have a new endpoint to add. Update the imports from the `./scim-types` file and declare a new route for resource types.

Reviewer: ```suggestion
We have a new endpoint to add. Update the imports from the `./scim-types` file and declare a new route for resource types in `entitlements.ts`:
```

at first I saw scim-types.ts and thought that was the file we were adding to. dumb mistake, easy fix.

    familyName: string;
  };
  emails?: {primary: boolean, value: string, type: string}[];
  displayName?: string;
  locale?: string;
  meta?: {
    resourceType: string;
  }
  externalId?: string;
  groups?: [];
  password?: string;
  active?: boolean;
  detail?: string;
  status?: number;
  roles?: {value: string, display: string}[];
}
```

The User SCIM schema defines `roles` property as a list of objects that may contain properties named `value` and `display`, among others. Okta uses these properties for role data.

### Update the SCIM add users call to include roles

The first route defined is the `POST /Users` route definition. You need to add roles when saving to the database. Find the comment

```js
// Create the User in the database
```

and update the database command and the as shown.

```ts
// Create the User in the database
const user = await prisma.user.create({
  data: {
    org : { connect: {id: ORG_ID}},
    name,
    email,
    password,
    externalId,
    active,
    roles: {
      connect: newUser.roles?.map(role => ({id: parseInt(role.value)})) || []
    }
  },
  include: {
    roles: true
  }
});

console.log('Account Created ID: ', user.id);
```

> One more place to update in the `POST /Users` call. We need to return the roles in the response. Right below the `console.log()` update the `userResponse` to

Reviewer: ```suggestion
One more place to update in the `POST /Users` call. We need to return the roles in the response. Below the line  `console.log('Account Created ID: ', user.id);`, update the `userResponse` to
```

easier to jump to the right spot with a reference that's unique in the file

```

We defined a new schema URN for the characteristic SCIM resource. Following naming conventions for extension schemas, we substituted our company name (Best Apps) and added the app's name (Todo app). The format looks like this

```
urn:<Company name>:scim:schemas:extension:<App name>:1.0:<Custom entitlement>
```

Right now, there's a custom TypeScript type for `RESOURCE_TYPES`. Since we'll have custom schemas as a resource type, update the code.

```ts
export type RESOURCE_TYPES = 'Role' | 'ResourceType' | 'Schema';
```

SCIM defines required and optional attributes to describe a schema resource. We'll define the interfaces for a schema resource. Add the following interfaces to the `scim-types.ts` file. I added mine after the other interfaces and before the URNs.

```ts
export interface ISchema {
  id: string;
  name?: string;
  description?: string;
  attributes: IAttribute[];
  meta: IMetadata;
}

export interface IAttribute {
  name: string;
  description: string;
  type: string;
  multiValued: boolean;
  required: boolean;
  caseExact: boolean;
  mutability: string;
  returned: string;
  uniqueness: string;
}
```

> Characteristic is a unique resource type because it's a new, custom type extending from an existing schema. We must explicitly show this relationship for consuming SCIM clients, like Okta. Find the `IResourceType` interface. We'll add a new optional property, `schemaExtensions` and inline the type definition.

Reviewer: ```suggestion
`Characteristic` is a unique resource type because it's a new, custom type extending from an existing schema. We must explicitly show this relationship for consuming SCIM clients, like Okta. Find the `IResourceType` interface. We'll add a new optional property, `schemaExtensions` and inline the type definition.
```

i read characteristic as the english word instead of the resource type without code formatting

```ts
export interface IResourceType {
  id?: string;
  schemas: string[];
  name: string;
  description?: string;
  endpoint: string;
  schema: string;
  schemaExtensions?: {schema: string, required: boolean}[];
  meta: IMetadata;
}
```

SCIM clients expect a list of schemas that you offer in the SCIM server. You might've guessed what that means. You must wrap all the schemas in a SCIM `ListResponse`. Find `IListResponse` and add `ISchema` as a supported type. The `IListResponse` interface changes to:

```ts
export interface IListResponse<T extends IScimResource | IResourceType | ISchema> {
  schemas: string[];
  totalResults: number;
  startIndex: number;
  itemsPerPage: number;
  Resources: T[];
}
```

> Finally, we define what a characteristic attribute looks like by adding the interface shown below.

Reviewer: I am confused by calling it a "characteristic attribute" here and then having it extend IScimResource instead of IAttribute. potential rephrase: 
 
```suggestion
Finally, we define what a `characteristic` looks like by adding the interface shown below.
```

```

At the end of the file, add the code to register the `/Schemas` and `/Charactertistics` routes to the API.

```ts
scimRoute.use('/Schemas', schemasRoute);
scimRoute.use('/Characteristics', characteristicsRoute);
```

Serve the API by running `npm run serve-api` in a terminal window. In a second terminal window, run `npx localtunnel --port 3333` to create a local tunnel for the API. Keep track of the tunnel URL.

Back in the [Okta Admin console](https://developer.okta.com/login/), navigate to **Applications** > **Applications** and open the SCIM with governance Okta app. Navigate to **Provisioning** > **Integration**. Press **Edit** and update the **Base URL** using the new tunnel URL. Don't forget to keep the `/scim/v2` at the end of the URL. The URL should look something like

```
https://{yourTunnelSubdomain}.loca.lt/scim/v2
```

Press **Save**.

Okta discovers schemas and resource types when updating the provisioning configuration. If you look at the HTTP call tracing in the terminal window serving the API, you'll see that Okta made a GET request to both `/Schemas` and `/Characteristics`.

> Navigate to the **Governance**. **Characteristic** may take 10-15 minutes to populate, but you'll see the display name and value when it does. Go **< Back to application** and navigate to **Assignments**. Open the user context menu for "Trinity" by pressing the three vertical dots icon menu and opening **View entitlements**. Press **Edit** and **Customize entitlements** to add the `is_tall` user characteristic. **Save** the changes and navigate back to the Okta SCIM app.

Reviewer: <img width="310" height="292" alt="image" src="https://github.com/user-attachments/assets/e5322b0d-2947-47cd-acbd-42f0a552bcab" />
:partying_face: it ended up working perfectly in an integrator account!
